_blogposts/compiler/string-literal-types-in-reason.mdx

```diff
@@ -0,0 +1,215 @@
+
+```reasonml
+type utf = [
+  | `utf8
+  | `utf19
+];
+type ascii = [
+  | `ascii
+]
+
+```
+
+The cool thing is that you can creat union types by simply put them together:
+
+```reasonml
+type encoding = [
+  | utf
+  | ascii
+]
+```
+
+The compiler even supports a sugar over named string literal types:
+
+```reasonml
+let classify = (enc) =>
+  switch (enc) {
+  | #utf => "utf" // string literals belong to utf
+  | #ascii => "asci" // string literals belog to ascii
+  };
+```
+
+The compiler would generate well optimized code as below:
+
+```js
+function classify(enc) {
+  if (enc === "ascii") {
+    return "asci";
+  } else {
+    return "utf";
+  }
+}
+```
+
+## String literal types in bindings
+
+Since string literal types are just strings after type checking, user can use it
+bind to js libs directly without any conversion as below:
+
+```reasonml
+type encoding = [
+  | `hex
+  | `utf8
+  | `ascii
+  | `latin1
+  | `base64
+  | `ucs2
+  | `base64
+  | `binary
+  | `utf16le
+];
+
+[@bs.val] [@bs.module "fs"]
+external readFileSync: (string, encoding) => string = "readFileSync";
+```
+## String literal types attached to data
+
+Since Reason is a typed language, you can not mix data of different types in a collection, for example, you will get a type error when writing code like this: `[ 3, "3" ]`.
+
+The deep reason is that if the compiler allows you to do such things, after you box different types of data in a single collection, it is hard to give such collection a type and process it later.
+
+With string literal types, you can do things like this:
+
+```reasonml
+[ 3 -> `Int , "3" -> `String ]
+```
+
+Note the generated code for `` 3 -> `Int`, "3"-> `String `` would be:
+
+```js
+{ NAME: "Int", VAL : 3}
+{ NAME: "String", VAL : "3"}
+```
+
+And you can also write code to process such collections:
+
+```reasonml
+let handle = (xs) =>
+  Belt.List.map(
+    xs,
+    fun
+    | `Int(n) => n
+    | `String(s) => String.length(s),
+  );
+```
+
+The generated code would be:
+
+```js
+function handle(xs) {
+  return Belt_List.map(xs, function (param) {
+    if (param.NAME === "Int") {
+      return param.VAL;
+    } else {
+      return param.VAL.length;
+    }
+  });
+}
+```
+To conclude, string literal types give users a convenient way to mix data with different types together and process it via pattern matching later.
+
+## Declaring types for string literal types attached to data
+
+Type inference is great during development, user can also write down the formal types for string literal types attched to data:
+
+```reasonml
+type number_or_string = [ 
+  | `Int(int) 
+  | `String(string)
+];
+```
+
+## Further reading
+
+Here we only cover some basic usage of string literal types, user can refer [here](https://caml.inria.fr/pub/docs/manual-ocaml/lablexamples.html#s%3Apolymorphic-variants) for more advanced stuff. The type theory is almost the same, however, we adapt it to make sure it's compiled into string literals to match the JS runtime.
```

Comment:
`user can refer` -> Maybe start a new sentence here. `For more advanced usage, see [here](https://...)`